facilitator/src/transport/gcs.rs

```diff
@@ -82,15 +122,26 @@ impl fmt::Debug for OauthTokenProvider {
             .finish()
     }
 }
+
 impl OauthTokenProvider {
     /// Creates a token provider which can impersonate the specified service
     /// account.
-    fn new(account_to_impersonate: Option<String>) -> OauthTokenProvider {
-        OauthTokenProvider {
+    fn new(
+        account_to_impersonate: Option<String>,
+        key_file_reader: Option<Box<dyn Read>>,
+    ) -> Result<OauthTokenProvider> {
+        let key_file: Option<ServiceAccountKeyFile> = match key_file_reader {
+            Some(reader) => {
+                serde_json::from_reader(reader).context("failed to deserialize JSON key file")?
+            }
+            None => None,
```

Comment:
Nit: this could be `key_file_reader.map( |reader| { ... } )`

Reply:
I'm not sure it can, because we would need to return a `Result<>` from the closure passed to `Option::map` (since `serde_json::from_reader` may fail), and I don't think you can do that from an `FnOnce`.

Reply:
Hm, good point about returning a result. I think you can (an `Fn` or `FnOnce` can return whatever you like), but type inference gets pretty shaky when it comes to closures, so you'd probably wind up having to annotate things to the extent it would be less clear instead of more. 👍🏻